pkg/protocol/engine/attestation/slotattestation/manager.go

```diff
@@ -0,0 +1,178 @@
+			cw := e.Storage.Settings().LatestCommitment().Commitment().CumulativeWeight
+			index := e.Storage.Settings().LatestCommitment().ID().Index()
+			m.Initialize(index, cw)
+		})
+
+		return m
+	})
+}
+
+func NewManager(attestationCommitmentOffset iotago.SlotIndex, bucketedStorage func(index iotago.SlotIndex) kvstore.KVStore, weightsProviderFunc func() *account.Accounts[iotago.AccountID, *iotago.AccountID]) *Manager {
+	m := &Manager{
+		attestationCommitmentOffset: attestationCommitmentOffset,
+		weightsProviderFunc:         weightsProviderFunc,
+		bucketedStorage:             bucketedStorage,
+		pendingAttestations:         memstorage.NewIndexedStorage[iotago.SlotIndex, iotago.AccountID, *iotago.Attestation](),
+	}
+	m.tracker = NewTracker(m.computeAttestationCommitmentOffset)
+	m.TriggerConstructed()
+
+	return m
+}
+
+func (m *Manager) Initialize(lastCommitmentIndex iotago.SlotIndex, lastCumulativeWeight uint64) {
+	m.commitmentMutex.Lock()
+	m.lastCommittedSlot = lastCommitmentIndex
+	m.lastCumulativeWeight = lastCumulativeWeight
+	m.commitmentMutex.Unlock()
+
+	m.TriggerInitialized()
+}
+
+func (m *Manager) Shutdown() {
+	m.TriggerStopped()
+}
+
+func (m *Manager) AddAttestationFromBlock(block *blocks.Block) {
+	m.commitmentMutex.RLock()
+	defer m.commitmentMutex.RUnlock()
+
+	// We only care about attestations that are newer than the last committed slot.
+	if block.ID().Index() <= m.lastCommittedSlot {
+		return
+	}
+
+	// Attestations that are older than m.lastCommittedSlot - m.attestationCommitmentOffset don't have any effect, so we ignore them.
+	if block.SlotCommitmentID().Index() <= m.computeAttestationCommitmentOffset() {
+		return
+	}
+
+	newAttestation := iotago.NewAttestation(block.Block())
+
+	storage := m.pendingAttestations.Get(block.ID().Index(), true)
+	storage.Compute(block.Block().IssuerID, func(currentValue *iotago.Attestation, exists bool) *iotago.Attestation {
+		if !exists {
+			return newAttestation
+		}
+
+		// Replace the stored attestation only if the new one is greater.
+		if newAttestation.Compare(currentValue) == 1 {
+			return newAttestation
+		}
+
+		return currentValue
+	})
+}
+
+func (m *Manager) Commit(index iotago.SlotIndex) (newCW uint64, attestationsMap *ads.Map[iotago.AccountID, iotago.Attestation, *iotago.AccountID, *iotago.Attestation], err error) {
+	m.commitmentMutex.Lock()
+	defer m.commitmentMutex.Unlock()
+
+	// Remove all pending attestations of slot and apply to tracker.
+	pendingAttestations := m.pendingAttestations.Evict(index)
+	if pendingAttestations != nil {
+		pendingAttestations.ForEach(func(key iotago.AccountID, value *iotago.Attestation) bool {
+			m.tracker.TrackAttestation(value)
+			return true
+		})
+	}
+
+	// Get all attestations for the valid time window of attestationSlotIndex up to index (as we just applied the pending attestations).
+	attestationSlotIndex := m.computeAttestationCommitmentOffset()
+	attestations := m.tracker.Attestations(attestationSlotIndex)
```

Comment:
`attestations` can be `nil` here and used in the range.

Reply:
This is valid and will just result in 0 iterations. Don't see an issue here.